File: leetcode/141_Linked_List_Cycle/solution.cpp
/*
 * 141. Linked List Cycle
 *
 * Given a linked list, determine if it has a cycle in it.
 *
 * Follow up:
 * Can you solve it without using extra space?
 * */

#include <iostream>
/*
 * Definition for singly-linked list.
 */
 struct ListNode {
      int val;
      ListNode *next;
      ListNode(int x) : val(x), next(NULL) {}
 };
 
class Solution {
	public:
		bool hasCycle(ListNode *head) {
			if(head == NULL || head->next == NULL) {
				return false;
			} 
				
			ListNode* fast = head;
			ListNode* slow = head;
			while(fast->next != NULL && fast->next->next != NULL) {
				fast = fast->next->next;
				slow = slow->next;
						
				if(slow == fast) {
					return true;
				}
			} 
				
			return false;
		}
};

int  main() {
	ListNode *a = new ListNode(1);
	ListNode *b = new ListNode(2);
	b->next = a;
	ListNode *c = new ListNode(3);
	c->next = b;
	ListNode *d = new ListNode(4);
	d->next = c;
	a->next = d;

	Solution *test = new Solution();
	std::cout << test->hasCycle(d) << std::endl;	

	return 0;	
}

File: leetcode/141_Linked_List_Cycle/Solution.java
import java.util.Set;
import java.util.HashSet;

public class Solution {
    public static class ListNode {
        int val;
        ListNode next;

        ListNode(int x) {
            val = x;
            next = null;
        }
    }

    public boolean hasCycle(ListNode head) {
        if (head == null || head.next == null)
            return false;
        
        Set<ListNode> visited = new HashSet<>();

        while (head != null) {
            if (!visited.contains(head)) {
                visited.add(head);
            } else {
                return true;
            }

            head = head.next;
        }

        return false;
    }

    public static void main(String[] args) {
        ListNode a1 = new ListNode(1);
        ListNode a2 = new ListNode(2);
        ListNode a3 = new ListNode(3);
        ListNode a4 = new ListNode(4);
        a1.next = a2;
        a2.next = a3;
        a3.next = a4;
        a4.next = a1;

        Solution test = new Solution();
        System.out.println(test.hasCycle(a1));     
    }
}